src/kademlia/get_item.cpp

```diff
@@ -70,6 +70,14 @@ void get_item::got_data(bdecode_node const& v,
 		{
 			if (!m_data.assign(v, salt, seq, pk, sig))
 				return;
+
+			// For get_item, we should call callback when we get data.
+			// The data maybe is not the latest, but we can update it later.
+			// Waitting transaction timeout (15 seconds) is not efficient.
+			// For put_item, it put new values to DHT when done(), so this
+			// change shouldn't break anything since put_item callback
+			// almost doing nothing, just a notification.
+			m_data_callback(m_data);
```

Comment:
well, the call you add in this patch is early/non-authoritative. The call to this callback that already exists in the code is the last one, and authoritative (meaning we've exhausted the search and likely timed out some nodes).

I think this is important. Even though most alerts are just unsolicited updated, some are responses to calls. This is the latter kind. It's conceivable and maybe even likely, that the client want to manage some form of resource between the call and the return. If you want to keep state for instance, it can be hard to know when it's OK to free the memory. (obviously you can tie it to the lifespan of the handler function itself, but there may be less trivial examples)